[IV-23-24] Objetivo 5

shvtwp opened the pull request:
* [x] ¿Se han establecido criterios previos de elección de la imagen base, así
como los criterios de búsqueda que han conducido a las imágenes que se están
considerando?
* [x] ¿Se ha tenido en cuenta que la imagen es para testear, y que por tanto
tiene una función específica, tal como testear cómo va nuestra aplicación en la
última versión publicada del lenguage?
* [x] ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?
* [x] ¿Se han tenido en cuenta criterios reales para la elección, y no criterios
irrelevantes como cuanto tiempo se tarda en construirla? (Se va a construir una
vez, se va a usar cientos o miles de veces)
* [x] ¿Se han seguido las buenas prácticas en las órdenes que construyen
   la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?
* [x] ¿La imagen incluye sólo y exclusivamente la infraestructura
   necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
   "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
   comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
   nuestro proyecto?
* [ ] ¿Se han documentado y enlazado los commits a las imágenes que se
   han testeado?
* [x] ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los
   tests que se van a lanzar?
* [ ] ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
en vez del directorio de trabajo?
* [x] ¿Se han establecido claramente los criterios de búsqueda y de aceptación de
   la imagen base?
* [ ] ¿Tienes claro cuales son las condiciones en las que tiene que actualizarse
la imagen en DockerHub y has configurado el workflow de acuerdo con ello? En
particular, ¿sabes qué condiciones tienen que darse para que se actualice y qué
condiciones *no* deben darse?
* [ ] Si se usan GitHub actions para su publicación, ¿se han considerado
cuidadosamene los criterios (ramas, paths) para que se lancen?
* [ ] ¿Se ha comprobado que el ENTRYPOINT del contenedor no escriba nada en el
directorio que monta, ya que no tiene derecho de escritura?
* [ ] ¿Se ha abierto un shell en el contenedor una vez construida la imagen y
comprobado que los ficheros que se necesitan están todos incluidos, y que no hay
ficheros que no se necesiten?

shvtwp committed:
Establece imagen base closes #30

shvtwp committed:
Añade dockerfile closes #31

Se ha creado el archivo de configuración del Docker

shvtwp committed:
Automatiza actualización del contenedor closes #33

Mediante el nuevo fichero, se automatiza la actualización del docker

shvtwp committed:
Actualiza README.md closes #32

Ahora incluye los comandos para ejecutar los test en el contenedor. Con esto se completa la configuración del entorno para pruebas closes #29

danieeeld2 commented on dockerfile:
```diff
@@ -0,0 +1,11 @@
+
+LABEL maintainer="shvtwp"
+
+WORKDIR /app/test
```
¿Aquí no necesitas actualizar y cagar ninguna dependencia? Si no tienes, entonces lo veo bien. En caso contrario, deberías copiar previamente el archivo que gestiona las dependencias y ejecutar las órdenes necesarias para actualizar e instalar dichas dependencias en la imagen

shvtwp replied:
Solucionado [b7c0fae](https://github.com/shvtwp/DePendiente/pull/34/commits/b7c0fae7915ec11f8507cc930752ad6606b400a7)

danieeeld2 reviewed (commented):
Lo que has hecho no lo veo mal, pero puedes tener el problema que, al haber tantos compañeros que ya han subido y superado el objetivo 5, te diga que la imagen `denoland/deno:alpine` ya no la puedes usar como solución. De todas formas puedes preguntarle a JJ antes de hacer cualquier cambio.
Ánimo con los objetivos y felices fiestas 😄

JJ commented:
> @JJ Hola! ¿Es aún válida la solución que he elegido?

Bueno, sería conveniente que exploraras otras alternativas, como dice @danieeeld2 . Recuerda que en la explicación mencioné que ibais a aprender más trabajando solo con sistemas operativos o imágenes no oficiales, en cualquier caso

shvtwp committed:
Cambia selección de imagen base

shvtwp committed:
Cambia imagen base de dockerfile

shvtwp commented:
@JJ listo para revisión

JJ commented on dockerfile:
```diff
@@ -0,0 +1,12 @@
+ARG DENO_VERSION=1.39.2
+
+FROM denoland/deno:bin-$DENO_VERSION AS deno
+FROM ubuntu:latest
```
Estás usando la construcción en dos fases, lo que es una buena idea. Pero creo que la imagen que estás usando en la segunda fase no es precisamente óptima.

shvtwp replied:
Solucionado en [facd63e](https://github.com/shvtwp/DePendiente/pull/34/commits/facd63e33cbc1c00ca4df0cc6b2f837be3b8a256)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,12 @@
+ARG DENO_VERSION=1.39.2
```
¿Has mirado para qué se usa en concreto `ARG`? ¿Crees que lo podrías evitar en este caso?

shvtwp replied:
Solucionado en [facd63e](https://github.com/shvtwp/DePendiente/pull/34/commits/facd63e33cbc1c00ca4df0cc6b2f837be3b8a256)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+WORKDIR /app/test
+
+ENTRYPOINT ["deno", "test", "--allow-read", "test/horario_test.ts"]
```
Aqui debes usar el task manager que hayas elegido. Para ello tendrás que copiar también aquí lo que necesites para ejecutarlo.

shvtwp replied:
Solucionado en [facd63e](https://github.com/shvtwp/DePendiente/pull/34/commits/facd63e33cbc1c00ca4df0cc6b2f837be3b8a256)

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,55 @@
+
+[Ubuntu](https://hub.docker.com/_/ubuntu) es otra opción a considerar
+debido a que no pesa demasiado (26.01 MB) y personalmente
+estoy más familiarizada con este sistema operativo.
```
La familiarización no es un criterio objetivo, me temo. Debes intentar usar solo razones técnicas, y Alpine es tan Linux como Ubuntu. Y si quieres algo a mitad de camino, versiones ligeras de Debian podrían ser una alternativa.
Si quieres aprender un poco más, en contenedores construidos en dos fases se usan también las llamadas `distroless`, pero posiblemente con Alpine ocupando 5 megas no sé si merecerá la pena.

shvtwp replied:
Solucionado en [60b9b10](https://github.com/shvtwp/DePendiente/pull/34/commits/60b9b104f2f3fabc240de29d0805be3512d64023)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+LABEL maintainer="shvtwp"
+
+COPY --from=deno /deno /usr/local/bin/deno
```
Seguramente necesitarás copiar algún fichero más de tu repo, ¿no?

shvtwp replied:
Solucionado en [facd63e](https://github.com/shvtwp/DePendiente/pull/34/commits/facd63e33cbc1c00ca4df0cc6b2f837be3b8a256)

JJ commented:
Primero, muchas gracias por marcar sólo lo que se aplique.

>     * [x]  ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?

Quizás deberías tener en cuenta que las imágnes `bin` tienen también etiquetas *semánticas*, es decir, que la 1.39 va a corresponder a la última 1.39.x que haya en cada momento.
 
> 
>     * [x]  ¿Se han seguido las buenas prácticas en las órdenes que construyen
>       la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?

¿Dónde has seguido la buena práctica de crear un usuario no privilegiado?

JJ reviewed (changes requested):
Muy bien por haber usado esta estrategia de construcción, pero el resto convendría que le dieras una vuelta. Ánimo, ya queda menos.

shvtwp committed:
Modifica Dockerfile #31

Tratando de cambiar la imagen base a Alpine, me estaba dando el mismo error todo el tiempo, así que he cambiado la estrategia de construcción según lo que he comentado en imagen_base.md

shvtwp committed:
Especifica el shell para drakefile closes #35

shvtwp committed:
Justifica la elección de nueva imagen base

shvtwp commented:
@JJ He hecho actualizaciones según lo comentado

JJ commented on dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+RUN apk add curl
+
+RUN adduser -D -u 1000 usuario
```
Una de las buenas prácticas es agrupar en una sola orden todas las que se hagan secuencialmente, sobre todo si se hacen con el mismo usuario.

shvtwp replied:
Solucionado en [a7eb5b5](https://github.com/shvtwp/DePendiente/pull/34/commits/a7eb5b5ac67bd982897f16163c510e522447c632)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,24 @@
+FROM frolvlad/alpine-glibc:alpine-3.18
```
Aparte de que estás usando una versión muy específica, con `glibc` en vez de `musl` que es la que trae por defecto, usar versiones específicas de Alpine tampoco parece que esté bien justificado.

shvtwp replied:
@JJ He intentado usar solo Alpine pero me da un error que no consigo solucionar, solo lo he conseguido con glibc que es, además, la imagen que usan en el [dockerfile](https://github.com/denoland/deno_docker/blob/main/alpine.dockerfile) de la imagen oficial de Deno basado en Alpine . Sigo intentando solucionarlo, pero no estoy teniendo mucho éxito.

JJ replied:
Normalmente la última disponible tiene convencionalmente `latest`. La versión que has incluido ni siquiera es la última de Alpine para ese contenedor en concreto... Mira en [el repo](https://hub.docker.com/r/frolvlad/alpine-glibc/tags) para ver las versiones disponibles. Y ánimo.

shvtwp replied:
@JJ También probé a usar la más reciente, pero por algún motivo se aborta antes de ejecutar los test cuando uso Alpine 3.19. La última versión con la que funciona es la 3.18. No sé muy bien qué debería hacer.

JJ replied:
Estás usando una versión con `glibc`... Estás añadiendo complejidad a Alpine, que ya de por sí va a ser complicado. En todo caso, que uses una versión específica aquí es la objeción menos importante. Yo me concentraría en las otras, y una vez solucionadas, ver si se puede solucionar el problema con el sistema base.
Recuerda que una vez que uses una imagen base con sistema operativo e instales deno sobre ella, puedes optar por cualquier otra imagen "oficial". El problema es la imagen oficial de deno, pero no la imagen oficial de un SO con deno instalado encima.

JJ commented on dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+RUN curl -fsSL https://deno.land/x/install/install.sh | sh
+ENV DENO_INSTALL="/home/usuario/.deno"
+ENV PATH="${DENO_INSTALL}/bin:${PATH}"
```
Igual que en el caso anterior, se trata de la misma orden en dos líneas consecutivas.

shvtwp replied:
Solucionado en [a7eb5b5](https://github.com/shvtwp/DePendiente/pull/34/commits/a7eb5b5ac67bd982897f16163c510e522447c632)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,24 @@
+ENV DENO_INSTALL="/home/usuario/.deno"
+ENV PATH="${DENO_INSTALL}/bin:${PATH}"
+
+LABEL maintainer="shvtwp"
```
Las buenas prácticas también aconsejan que se ponga un número de versión.

shvtwp replied:
Solucionado en [fa025ec](https://github.com/shvtwp/DePendiente/pull/34/commits/fa025ece8211db31c157d30494723dbdb36d977e)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+LABEL maintainer="shvtwp"
+
+COPY ./drakefile.ts .
```
¿Seguro que no hay alguna cosa más que tendrás que copiar en deno?

JJ replied:
Aparte, también sería conveniente que configuraras deno para que drake esté descargado *dentro del contenedor*. Para eso tienes que crear un fichero con las dependencias, que tendrías que copiar también y ejecutar.

shvtwp replied:
Solucionado en [b7c0fae](https://github.com/shvtwp/DePendiente/pull/34/commits/b7c0fae7915ec11f8507cc930752ad6606b400a7)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+WORKDIR /app
+
+ENTRYPOINT ["deno", "run", "-A", "drakefile.ts", "testHorario"]
```
¿`drake` no tiene un comando propio para funcionar?

JJ replied:
Pues parece que no...

JJ replied:
Este fichero ¿en qué directorio estará? ¿Tendrá algo que ver con el fichero `drakefile` que has copiado?

JJ replied:
La tarea para test es siempre `test`, no `testHorario`

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,57 @@
+
+## Conclusión
+
+Finalmente, me he decantado por elegir Alpine e
```
Pero no has usado la imagen oficial de Alpine, ¿no?

JJ replied:
La presentación de las diferentes imágenes deberías también hacerla, con los criterios para elegir esta específica.

shvtwp replied:
Solucionado en [777c6a0](https://github.com/shvtwp/DePendiente/pull/34/commits/777c6a0a0e872edbeff44a6a49ed9696730f46ce)

JJ reviewed (changes requested):
En general, es conveniente que modifiquéis incrementalmente lo que hayáis hecho en vez de empezar desde cero. La imagen que usabas antes no tenía nada malo en si, sólo que podías haberle modificado alguna cosa en el mismo sentido que lo has hecho aquí. En este caso puedes hacer lo que veas, pero la imagen que has presentado tampoco está bien justificada, aparentemente se trata de una justificación genérica y lo específica de la de frolvlad.

JJ reviewed (commented):
Algunos comentarios adicionales.

shvtwp committed:
Agrupa órdenes de dockerfile

shvtwp committed:
Añade versión

shvtwp committed:
Añade archivo gestor de dependencias closes #36

shvtwp committed:
Añade archivo gestor de dependencias closes #36

shvtwp commented:
@JJ He solucionado todo excepto lo de Alpine. He intentado hacerlo de varias formas, pero siempre resulta en el mismo [error](https://github.com/shvtwp/DePendiente/assets/127195032/97e4880b-a6ef-4b0f-8296-56ccb8ba9d5e) y me quedo sin opciones. Quizá lo mejor sería volver a la imagen oficial de Deno, ya que la actual no funciona tampoco con la última versión de Alpine.

JJ commented:
> @JJ He solucionado todo excepto lo de Alpine. He intentado hacerlo de varias formas, pero siempre resulta en el mismo [error](https://github.com/shvtwp/DePendiente/assets/127195032/97e4880b-a6ef-4b0f-8296-56ccb8ba9d5e) y me quedo sin opciones. Quizá lo mejor sería volver a la imagen oficial de Deno, ya que la actual no funciona tampoco con la última versión de Alpine.

El error no parece tener nada que ver con el hecho de que uses una máquina base u otra...

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+LABEL maintainer="shvtwp" \
+      version="5.0.2"
+
+COPY ./drakefile.ts ./deno.json ./
```
¿No habrá que copiar también los lockfiles?
De hecho, *después* de copiar tanto deno.json como el lockfile tienes que descargar los ficheros correspondientes en la caché.

JJ replied:
Aparte, ¿en qué directorio estás copiando esto? ¿Está en un directorio donde luego lo pueda encontrar deno?

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+USER usuario
+
+RUN curl -fsSL https://deno.land/x/install/install.sh | sh
```
Si instalas algo que sólo vas a usar en la construcción, ¿no será conveniente borrarlo luego?
En todo caso, Alpine trae instalado `wget` que hace la misma función y no haría falta instalar nada.

JJ replied:
Cuando haces esto también te da un mensaje:

```
Manually add the directory to your $HOME/.bashrc (or similar)
  export DENO_INSTALL="/home/usuario/.deno"
  export PATH="$DENO_INSTALL/bin:$PATH"
```
Una parte la estás haciendo abajo. Pero te falta la otra parte, ¿no?

shvtwp replied:
Solucionado en [6b5cce5](https://github.com/shvtwp/DePendiente/pull/34/commits/6b5cce52e49439f743f61920ef76f5d1ecdc1a49)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM frolvlad/alpine-glibc:alpine-3.18_glibc-2.34
```
Vuelvo a insistir: no tienes por qué usar esta versión específica de Alpine. Hay otras opciones para Alpine, incluso la oficial, que posiblemente no te den ningún problema.

JJ reviewed (changes requested):
Has mejorado algo la construcción de la imagen, pero todavía sigues sin exponer los criterios que te han llevado a elegir esta específica.

JJ commented:
Efectivamente:

```
Could not initialize cache database '/app/.cache/deno/dep_analysis_cache_v1', deleting and retrying... (unable to open database file: /app/.cache/deno/dep_analysis_cache_v1

Caused by:
    Error code 14: Unable to open the database file)
Failed to open cache file '/app/.cache/deno/dep_analysis_cache_v1', opening in-memory cache.
Could not initialize cache database '/app/.cache/deno/node_analysis_cache_v1', deleting and retrying... (unable to open database file: /app/.cache/deno/node_analysis_cache_v1

Caused by:
    Error code 14: Unable to open the database file)
Failed to open cache file '/app/.cache/deno/node_analysis_cache_v1', opening in-memory cache.
error: Module not found "file:///app/drakefile.ts".
```

Deberías ejecutar las imágenes creadas cada vez que las crees, a ver si funcionan...

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+COPY ./drakefile.ts ./deno.json ./
+
+WORKDIR /app
```
El `WORKDIR` tiene que ser `/app/test`

shvtwp replied:
Solucionado en [6b5cce5](https://github.com/shvtwp/DePendiente/pull/34/commits/6b5cce52e49439f743f61920ef76f5d1ecdc1a49)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM frolvlad/alpine-glibc:alpine-3.18_glibc-2.34
+
+WORKDIR /
```
Este es un directorio privilegiado en el que no se debería trabajar.

shvtwp replied:
Solucionado en [6b5cce5](https://github.com/shvtwp/DePendiente/pull/34/commits/6b5cce52e49439f743f61920ef76f5d1ecdc1a49)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+RUN apk update && apk upgrade && \
+    apk add curl && adduser -D -u 1000 usuario
+
+USER usuario
```
A partir de aquí estás trabajando con un  usuario no privilegiado en un directorio con privilegios. Deberías crear un directorio sin privilegios (que ya el guión te dice claramente cuál es) dándole permisos a este usuario para que trabaje en él.

shvtwp replied:
Solucionado en [6b5cce5](https://github.com/shvtwp/DePendiente/pull/34/commits/6b5cce52e49439f743f61920ef76f5d1ecdc1a49)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+WORKDIR /
+
+RUN apk update && apk upgrade && \
```
Supongo que el `update` y el `upgrade` lo haces porque tienes que instalar curl; pero eso va a provocar que la imagen resultante ocupe mucho, porque se descargan todas las cachés. Como te he dicho, Alpine tiene por omisión instalado `wget`  que hace la misma función (aunque no necesariamente con la misma sintaxis)

shvtwp replied:
Solucionado en [6b5cce5](https://github.com/shvtwp/DePendiente/pull/34/commits/6b5cce52e49439f743f61920ef76f5d1ecdc1a49)

JJ commented:
> @JJ He solucionado todo excepto lo de Alpine. He intentado hacerlo de varias formas, pero siempre resulta en el mismo [error](https://github.com/shvtwp/DePendiente/assets/127195032/97e4880b-a6ef-4b0f-8296-56ccb8ba9d5e) y me quedo sin opciones. Quizá lo mejor sería volver a la imagen oficial de Deno, ya que la actual no funciona tampoco con la última versión de Alpine.

Además, tienes que montar el directorio local en `/app/test`, no en `app` como dice ese comando.

JJ commented:
OK, ya me parece que estoy encontrando el problema por el que tienes que usar glibc. Se debe a [esto](https://stackoverflow.com/a/66974607/891440).
Solución: no usar Alpine, sino alguna versión ligera de Debian, o usar Alpine con glibc

JJ commented:
Por recapitular un poco, va a haber que hacer una serie de retoques sobre lo que has hecho. Si quieres hacemos una tutoría la semana que viene, pero por recapitular un poco, creo que podrías seguirlo tú misma

1. Es mejor que *no* uses Alpine con deno. Sólo funciona si usas la opción glibc, para lo que tienes que usar esa específica. Todo esto es porque, como expliqué en clase, Alpine lleva `musl`, lo que provoca que si instalas `deno`, simplemente te diga que no lo encuentra aunque esté en el PATH. Solución: usar alguna versión ligera de Ubuntu o de Debian. Examina diferentes opciones que tengas (incluyendo las oficiales) y elige la que veas más conveniente entre las disponibles
2. `deno` usa una caché para descargarse los ficheros que usa, *aunque uses deno.json* aparentemente. La caché, por omisión, la pone en `/.cache`. Solución: crea ese directorio con permisos de escritura y lectura y ejecución para todo el mundo
3. La versión que elijas de Debian y Ubuntu puede que no tenga ni `curl` ni, por supuesto, `zip`, instalado. Solución: instálalo con el gestor de paquetes (y bórralos cuando te descargues deno)

Una vez hecho esto, pruébalo sobre tus fuentes y comprueba que no hayas cometido ninguno de los errores comunes. Si aún así no va, vemos la posibilidad de hacer una tutoría la semana que viene si quieres.

JJ reviewed (commented):
Mira estos otros comentarios que te he hecho, por favor, aunque la mayoría se resolverán solos una vez que sigas lo que te he comentado antes.

shvtwp committed:
Soluciona error en drakefile

testHorario --> test

shvtwp committed:
Actualiza justificación de imagen base #30

shvtwp committed:
Actualiza Dockerfile #31

Se modifica el Dockerfile según los cambios en la selección de la imagen. Se procura que la imagen sea lo más pequeña posible, por lo que se ha cambiado también la estrategia de construcción para minimizar el tamaño. También se solucionan los cambios sugeridos por el profesor

shvtwp committed:
Actualiza dockerfile según errores frecuentes

En el archivo de errores frecuentes dice que no se debe poner nada dentro de /app/test

shvtwp commented:
@JJ Solicito revisión. No he resuelto dos comentarios pero es porque no estoy segura de haberlos entendido.

JJ commented on dockerfile:
```diff
@@ -0,0 +1,39 @@
+RUN useradd -m -u 1000 usuario
+
+COPY --from=builder /home/usuario/.deno /home/usuario/.deno
+COPY --from=builder /.cache /.cache
```
Realmente en vez de copiarlo deberías crearlo en esta etapa. No creo que haya nada ahí.
En todo caso, como te digo en el comentario general, no parece que usar dos etapas te esté aportando mucho aquí.

shvtwp replied:
Solucionado en [893298d](https://github.com/shvtwp/DePendiente/pull/34/commits/893298d59dc196ce642849fdbe294e032f45776b)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,39 @@
+COPY --from=builder /.cache /.cache
+COPY --from=builder /app /app
+
+USER usuario
```
No sé si esto funcionará. Estás construyendo en dos etapas, y el usuario lo has creado en la etapa anterior. Al introducir otra etapa nueva, es una imagen totalmente nueva, que ya no tendrá el usuario.
En todo caso, ¿te parece que usar dos etapas aquí aporte algo? Me da la impresión de que vas a evitar tener todos los repositorios actualizados (al hacer apt upgrade) y poco más. En todo caso, usando la imagen Ubuntu va a ocupar bastante. ¿Por qué no consideras una imagen reducida, de Ubuntu o de Debian?

shvtwp replied:
Solucionado en [893298d](https://github.com/shvtwp/DePendiente/pull/34/commits/893298d59dc196ce642849fdbe294e032f45776b)

JJ commented on dockerfile:
```diff
@@ -0,0 +1,39 @@
+LABEL maintainer="shvtwp" \
+      version="5.0.2"
+
+COPY ./drakefile.ts ./deno.json ./deno.lock ./
```
Como te he dicho antes, cuando ejecutas el ENTRYPOINT el drakefile.ts que va a tomar va a ser el que "monte" desde el directorio externo. Y deno.json y deno.lock no los estás usando para nada. Podías usarlos para ejecutar `deno cache`, pero al hacer `deno run` lo va a ejecutar de todas formas. Si tienes dudas, ejecuta el contenedor para que pase los tests con esa sentencia y sin ella, a ver si hay alguna diferencia. Posiblemente no haya ninguna.
En general, las mejores prácticas incluyen meter en una imagen solo y exclusivamente lo que sepamos que funciona, y como funciona. Meter comandos adicionales porque los sugiera CoPilot (que no creo que sea el caso) o los hayas visto en algún lado no conduce a un buen aprendizaje del funcionamiento de una imagen.

shvtwp replied:
Solucionado en [893298d](https://github.com/shvtwp/DePendiente/pull/34/commits/893298d59dc196ce642849fdbe294e032f45776b)

JJ reviewed (changes requested):
Siempre es mejor estrategia, como te dije, tratar de solucionar los problemas que se presenten, en vez de comenzar desde cero, porque se introducen nuevos problemas que, a su vez, hay que solucionar.

Partiendo de lo que tienes aquí, *sin empezar desde cero de nuevo, por favor*

1. Considera usar como imagen base alguna más ligera. Con ubuntu es complicado, pero puedes usar Debian que usa el mismo formato de paquetes y todo va a ser más familiar. Las imágenes Debian ligeras se llaman `slim`
2. No uses dos fases de construcción a menos que estés muy, muy segura de lo que estás haciendo, y hazlo solo si simplifican el resultado, no si lo complican por un ahorro de tamaño cuestionable.
3. Si finalmente usas Ubuntu (tampoco pasa nada), es posible que ya estén instalados tanto curl como unzip y por tanto no tengas que usar ni apt-get update ni apt-get upgrade (con el consiguiente aumento de tamaño)
4. En ese caso, ya no tienes que  copiar de la imagen base, porque no necesitas ahorrar ningún tamaño. También puedes borrar `curl` y `unzip` si ya no lo vas a usar.

La mayoría de los comentarios se resuelven solos si haces lo de arriba. En todo caso, te recomiendo encarecidamente que ejecutes los contenedores una vez construidos, porque así se ve automáticamente los problemas que hay.

JJ commented on .github/workflows/autom.yml:
```diff
@@ -0,0 +1,27 @@
+  push:
+    paths:
+      - 'Dockerfile'
+      - 'deno.lock'
```
Adicionalmente, si realmente no vas a copiar estos ficheros a la imagen, no parece que sea necesario que se relance cuando cambien; pero si van a copiarse, tendrás que poner aquí *todos* los que se vayan a incluir en la imagen, incluyendo `drakefile.ts`. `paths` aquí indica cuales son los ficheros que van a provocar que se relance este action, reconstruyendo la imagen.

shvtwp replied:
Solucionado en [22bff4c](https://github.com/shvtwp/DePendiente/pull/34/commits/22bff4c0c2f86bb848d36c39f9815043e6f9378e)

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,72 @@
+de la imagen sería 223MB, que equivale a los 72MB
+que pesa Ubuntu más que Alpine.
+Si además usamos una estrategia de construcción
+multicapa, podemos reducir el tamaño a 215MB.
```
OK, bien justificado el uso de dos capas, pero sólo si efectivamente funciona, que me parece que no es el caso... Realmente puedes ahorrarte eso si no lanzas el `update` y `upgrade`, una vez que compruebes que `curl` y `zip` están instalados (`curl` es muy proable, `zip` también). Y si no están instalados, simplemente no hagas `upgrade` sino solo update, y borra tanto `curl` como `zip` una vez que esté instalado deno. Con eso deberías ahorrar los megas de sobra.

shvtwp replied:
Intento hacerlo así y en lugar de ahorrar megas, se le suma uno más si trato de desinstalarlos.

JJ replied:
También es verdaad... Gracias por probarlo. Dependiendo de la versión de docker build que uses, a veces sí lo reduce, pero no siempre, creo que tienes que usar buildx. En fin, da igual. Buen trabajo de todas formas.

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,72 @@
+[Alpine con glibc preinstalado](https://hub.docker.com/r/frolvlad/alpine-glibc), lo que también puede resultar problemático al
+tratarse de una versión muy específica.
+
+[Debian](https://hub.docker.com/_/debian) el principal problema
```
¿Has mirado las versiones `slim`?

shvtwp replied:
Solucionado en [91fab72](https://github.com/shvtwp/DePendiente/pull/34/commits/91fab72cdf63953cc70fc8c30f4ce88911603560)

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,72 @@
+posible instalar Deno en la imagen oficial, por
+lo que la alternativa sería usar la imagen
+[Alpine con glibc preinstalado](https://hub.docker.com/r/frolvlad/alpine-glibc), lo que también puede resultar problemático al
+tratarse de una versión muy específica.
```
Bueno, es la que usa `denoland/deno`... La mantendrán por la cuenta que le trae. La cuestión es que una vez que instales glibc, y luego curl y unzip, no creo que haya tantísima diferencia...

shvtwp replied:
En la versión de glibc preinstalado no necesito instalar curl y unzip, pero no puedo tampoco usar la última versión porque da error. Aún así, la diferencia de tamaño es notoria, mientras que esta pesa unos 160MB, una versión ligera de Debian sin dividir la construcción en etapas, pesa aproximadamente 246MB.

JJ replied:
Si no se puede usar la última versión, es un *red flag*, siempre se debería poder usar la última versión qeu es la única que corrige las vulnerabilidades.

shvtwp replied:
Resuelto en [91fab72](https://github.com/shvtwp/DePendiente/pull/34/commits/91fab72cdf63953cc70fc8c30f4ce88911603560)

JJ reviewed (changes requested):
OK, ya he visto la justificación, es buena, pero la construcción en dos fases complica las cosas así que no sé si merece la pena. Eficiencia frente a simplicidad, siempre se debe optar por la simplicidad.

shvtwp committed:
Soluciona errores en imagen_base.md

Había tamaños que no estaban bien, además, he elegido ahora Debian slim

shvtwp committed:
Elimina construcción en dos etapas #31

Además, realiza otros arreglos según lo comentado

shvtwp committed:
Actualiza el fichero de auromatización

shvtwp commented:
@JJ listo para revisión

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,78 @@
+de la imagen sería 269MB.
+Si además usamos una estrategia de construcción
+multicapa, podemos reducir el tamaño a 215MB.
+Según he [leído](https://canonical.com/blog/minimal-ubuntu-released), la imagen oficial de Ubuntu ya
```
Pero al final estás usando Debian...

shvtwp replied:
Sí, lo indico en la conclusión. Debian es más pequeña aunque sea por poco.

JJ commented on docs/imagen_base.md:
```diff
@@ -0,0 +1,78 @@
+el sistema operativo Alpine, que es con el que cuentan
+la mayoría de imágenes de DockerHub. Esta pesa 151MB.
+
+## Sistema operativo + Deno
```
Tampoco usan deno ya...

shvtwp replied:
Esto no lo he entendido, sigo instalando deno en la imagen base.

JJ replied:
Ya, pero lo que describes aquí son simplemente sistemas operativos donde se puede (o no, como en el caso de Alpine normal) instalar deno.

shvtwp replied:
Resuelto en [067737d](https://github.com/shvtwp/DePendiente/pull/34/commits/067737d14ed9db77d0304381ba43c03f73526db0)

JJ reviewed (approved):
:+1: 

Voy a darlo por bueno, pero otra vez la justificación está dessincronizada con lo que efectivamente has usado; por favor, corrígelo antes de fusionarlo o en todo caso antes de la presentación.

shvtwp committed:
Corrige imagen_base.md

No es correcto poner Sistema operativo + Deno, ya que también menciono Alpine y no se puede instalar Deno en esa imagen base
